### PR TITLE
Implement structs, FFI and GPU Driver

### PR DESCRIPTION
**Pull request by GlowingScrewdriver**
This PR introduces structs, exposing and building on LLVM's struct type.

Will bring #13 closer to completion

### CODE REVIEW

**Comment by GlowingScrewdriver** [comment]
Accidentally included a whole lot of commits from another branch here; hence I had to rebase and force-push

**Comment by GlowingScrewdriver** [comment]
Current design choices for Brilisp structs:
* No element names, only indices (as with LLVM)
* Structs will be allocated in memory using alloc and accessed through pointers only, just like arrays. So a struct is essentially an array of 1 struct.
* No nested structs
* Struct type definitions can come only in the global scope, alongside functions.

**Comment by GlowingScrewdriver** [comment]
Added a new test, `time`.
This program populates a struct with time fields and it passes to C standard library function `asctime`, which then pretty-prints it.
The fact that it works means we have interoperability with C API!

File: src/backend/c-lisp.py [hunk 1]
```diff
@@ -401,6 +422,24 @@ def gen_alloc_expr(self, expr, res_sym):
             ["set", [res_sym, ptr_type], ["alloc", size_sym]],
         ]
 
+    def is_struct_memb_expr(self, expr):
+        return expr[0] == "struct-memb"
```
Comment (chsasank) [inline src/backend/c-lisp.py:426]: call it `sptradd`

**Comment by chsasank** [comment]
@GlowingScrewdriver: please resolve the merge conflict before adding more code.

File: src/backend/tests/kernelprog/vecadd/vecadd.driver.sexp [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+        (declare device int) (set device 0)
+        (declare context (ptr int8)) (set context nullptr)
+        (declare module (ptr int8)) (set module nullptr)
+        (declare kernel_func (ptr int8)) (set kernel_func nullptr)
+
+        ;; CUDA initialization and context creation
+        ;; TODO [macro]: Wrap API calls with error-checking macros
+        (call cuInit 0)
+        (call cuDeviceGetCount (ptr-to devCount))
+        (call cuDeviceGet (ptr-to device) 0)
```
Comment (chsasank) [inline src/backend/tests/kernelprog/vecadd/vecadd.driver.sexp:78]: ptr-to should work without a set

File: src/backend/tests/kernelprog/vecadd/vecadd.driver.sexp [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+    ;; [WIP]
+    (define ((main void))
+
+        (declare i int)
+        (declare nullptr (ptr int8)) (set nullptr (inttoptr 0 (ptr int8)))
+
+        (declare devCount int) (set devCount 0)
+        (declare device int) (set device 0)
+        (declare context (ptr int8)) (set context nullptr)
+        (declare module (ptr int8)) (set module nullptr)
```
Comment (chsasank) [inline src/backend/tests/kernelprog/vecadd/vecadd.driver.sexp:71]: use a macro here something like ptr-void

Comment (chsasank) [reply src/backend/tests/kernelprog/vecadd/vecadd.driver.sexp:71]: also for cudevice

File: src/backend/tests/kernelprog/vecadd/vecadd.driver.sexp [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+    (define ((exit void) (status int)))
+    (define ((malloc (ptr float)) (sz int)))
+    (define ((rand float)))
+    ; From <math.h> (-lm)
+    (define ((fabsf float) (a float)))
+    (define ((fmaxf float) (a float) (b float)))
+    (define ((free void) (p (ptr float))))
+
+    ; Signatures from Numba's CUDA driver
+    ,@(get_cuda_signatures
```
Comment (chsasank) [inline src/backend/tests/kernelprog/vecadd/vecadd.driver.sexp:23]: make it get all signatures

File: src/backend/tests/kernelprog/vecadd/vecadd.driver.sexp (hunk 2 above)
Comment (GlowingScrewdriver) [reply src/backend/tests/kernelprog/vecadd/vecadd.driver.sexp]: Done

File: src/backend/tests/kernelprog/vecadd/vecadd_macros.py [hunk 5]
```diff
@@ -0,0 +1,123 @@
+# from numba_drvapi import (
```
Comment (chsasank) [inline src/backend/tests/kernelprog/vecadd/vecadd_macros.py:1]: remove all comments

File: src/backend/tests/kernelprog/vecadd/vecadd.driver.c [hunk 6]
```diff
@@ -35,12 +35,12 @@ void read_module(char * buf) {
 
 int main (int argc, char ** argv) {
     int devCount;
-    CUdevice device;
-    CUcontext context;
-    CUmodule module;
+    int device;
```
Comment (chsasank) [inline src/backend/tests/kernelprog/vecadd/vecadd.driver.c:38]: restore to originals

File: src/backend/tests/kernelprog/vecadd/vecadd_macros.py [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+        if memb.startswith("cu_"):
+            # Dynamicall create the macro variable
+            globe[memb] = get_clisp_type(getattr(numba_drvapi, memb))
+
+    return signatures
+
+def get_clisp_type(c_type):
+    """ Map a ctypes object to a C-Lisp type """
+    ptr_int8 = ["ptr", "int8"]
+    type_map = {
```
Comment (chsasank) [inline src/backend/tests/kernelprog/vecadd/vecadd_macros.py:65]: I think we can do a better job here. This looks super hacky.

File: src/backend/tests/kernelprog/vecadd/vecadd.driver.sexp [hunk 8]
```diff
@@ -0,0 +1,146 @@
...
+        (for ((set i 0)
+              (lt i N)
+              (set i (add i 1)))
+            (store (ptradd res i)
+                   (fadd (load (ptradd a i))
+                         (load (ptradd b i))))))
+
+    (define ((error_check void) (res int) (call_str (ptr int8)))
+        (if (ne res 0)
+            (,(printf call_str)
```
Comment (chsasank) [inline src/backend/tests/kernelprog/vecadd/vecadd.driver.sexp:29]: haha printf is now a proper macro!

Comment (GlowingScrewdriver) [reply src/backend/tests/kernelprog/vecadd/vecadd.driver.sexp:29]: Yes! We don't even need our old `print` utility now.

Comment (chsasank) [reply src/backend/tests/kernelprog/vecadd/vecadd.driver.sexp:29]: yayyy

File: src/backend/tests/kernelprog/vecadd/vecadd_macros.py (hunk 5 above)
Comment (GlowingScrewdriver) [reply src/backend/tests/kernelprog/vecadd/vecadd_macros.py]: Done

File: src/backend/tests/kernelprog/vecadd/vecadd.driver.c (hunk 6 above)
Comment (GlowingScrewdriver) [reply src/backend/tests/kernelprog/vecadd/vecadd.driver.c]: Done

File: src/backend/tests/kernelprog/vecadd/vecadd_macros.py (hunk 7 above)
Comment (GlowingScrewdriver) [reply src/backend/tests/kernelprog/vecadd/vecadd_macros.py:39]: I've gone this length for 2 reasons:
* I don't want to inject any code into `numba_drvapi.py`. I want to use it as-is, no modifications, as just another import.
* ctypes seems to create classes on-the-fly. In some cases it is difficult to compare the classes of ctypes objects (e.g. using `isinstance`)

Consider the following Python session:
```python
>>> import numba_drvapi
>>> numba_drvapi.cu_device
<class 'ctypes.c_int'>
>>> import ctypes
>>> isinstance(numba_drvapi.cu_device, ctypes.c_int)
False
```

File: src/backend/c-lisp.py [hunk 9]
```diff
@@ -528,16 +543,52 @@ def compile(self, expr):
...
         return ExpressionResult(instrs, res_sym, ptr_type)
 
 
+class StructPtrAddExpression(Expression):
+    @classmethod
+    def is_valid_expr(cls, expr):
+        return expr[0] == "sptradd"
+
+    def is_struct_pointer_type(self, typ):
+        return (
```
Comment (GlowingScrewdriver) [inline src/backend/c-lisp.py:561]: consider using
```python
verify_shape(typ, ["ptr", ["struct", None]])
and typ[1][1] in self.ctx.struct_types
```
if this is too complicated.

**Comment by chsasank** [comment]
Finally, ready to merge!

**Comment by GlowingScrewdriver** [comment]
LGTM

**Pull request merged**